packages/agent/src/agent/http/index.ts

```diff
@@ -424,63 +435,171 @@ export class HttpAgent implements Agent {
+          status: response.status,
+          statusText: response.statusText,
+          headers: httpHeadersTransform(response.headers),
+        },
+        requestId,
+      };
     };
+    const queryPromise = new Promise<ApiQueryResponse>((resolve, reject) => {
+      makeQuery()
+        .then(response => {
+          resolve(response);
+        })
+        .catch(error => {
+          reject(error);
+        });
+    });
 
-    // TODO: remove this any. This can be a Signed or UnSigned request.
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    let transformedRequest: any = await this._transform({
-      request: {
-        method: 'POST',
-        headers: {
-          'Content-Type': 'application/cbor',
-          ...(this._credentials ? { Authorization: 'Basic ' + btoa(this._credentials) } : {}),
-        },
-      },
-      endpoint: Endpoint.Query,
-      body: request,
+    const subnetStatusPromise = new Promise<SubnetStatus | void>((resolve, reject) => {
+      if (!this.#verifyQuerySignatures) {
+        resolve(undefined);
+      }
+      const subnetStatus = this.#subnetKeys.get(canisterId.toString());
+      if (subnetStatus) {
+        resolve(subnetStatus);
+      } else {
+        this.fetchSubnetKeys(canisterId)
+          .then(response => {
+            resolve(response);
+          })
+          .catch(error => {
+            reject(error);
+          });
+      }
     });
+    const [query, subnetStatus] = await Promise.all([queryPromise, subnetStatusPromise]);
+    // Skip verification if the user has disabled it
+    if (!this.#verifyQuerySignatures) {
+      return query;
+    }
+    return this.#verifyQueryResponse(query, subnetStatus);
+  }
 
-    // Apply transform for identity.
-    transformedRequest = await id?.transformRequest(transformedRequest);
+  /**
+   * See https://internetcomputer.org/docs/current/references/ic-interface-spec/#http-query for details on validation
+   * @param queryResponse - The response from the query
+   * @param subnetStatus - The subnet status, including all node keys
+   * @returns ApiQueryResponse
+   */
+  #verifyQueryResponse = (
+    queryResponse: ApiQueryResponse,
+    subnetStatus: SubnetStatus | void,
+  ): ApiQueryResponse => {
+    if (!subnetStatus || this.#verifyQuerySignatures === false) {
+      // This should not be called if the user has disabled verification
+      return queryResponse;
+    }
+    const { status, signatures, requestId } = queryResponse;
+
+    const domainSeparator = new TextEncoder().encode('\x0Bic-response');
+    signatures?.forEach(sig => {
+      const { timestamp, identity } = sig;
+      const nodeId = Principal.fromUint8Array(identity).toText();
+      let hash: ArrayBuffer;
+
+      // Hash is constructed differently depending on the status
+      if (status === 'replied') {
+        const { reply } = queryResponse;
+        hash = hashOfMap({
+          status: status,
+          reply: hashOfMap(reply),
```

Comment:
According to the latest specification and the final clarification on the `hash_of_map`, for nested map, we don't hash them twice. So instead of using `hashOfMap(reply)`, we just put `reply` here and adapt the `hashValue` function to support Record<string , unknown>  type and call `hash_of_map` there.

Reply:
Resolved. Hope this is the last adjustment!